tinygrad/lazy.py

```diff
@@ -198,89 +164,120 @@ def cast(self:LazyBuffer, arg:DType) -> LazyBuffer: return elementwise_op(UnaryO
-      if root.st.contiguous and root != self and prod(ret.st.shape) == prod(root.shape):
-        return root.movement_op(MovementOps.RESHAPE, ret.st.shape)
-
-    return ret
+  def expand_op(self: LazyBuffer, arg:Tuple[int, ...]) -> LazyBuffer:
+    if self.shape == arg: return self
+    if not self.realized and self.op.op == MovementOps.EXPAND:
+      self.op.src[0].children.discard(self)
+      return self.op.src[0].expand_op(arg)
+    ret = create_lazybuffer(self.device, ShapeTracker(self.st).expand(arg), MovementOps, LazyOp(MovementOps.EXPAND, (self,), arg, (self,)), self.dtype)
+    return self.prune_movement_nops(ret)
 
+  def permute_op(self: LazyBuffer, arg:Tuple[int, ...]) -> LazyBuffer:
+    if arg == tuple(range(len(self.shape))): return self
+    if not self.realized and self.op.op == MovementOps.PERMUTE:
+      self.op.src[0].children.discard(self)
+      return self.op.src[0].permute_op(tuple([self.op.arg[i] for i in arg]))
+    if not self.realized:
+      if PUSH_PERMUTES and self.optype == ReduceOps:
+        # reduceops have one buffer input, permute it
+        narg = tuple([self.op.arg[arg[i]] for i in range(len(arg))])
+        src, rop = self.op.src[0], self.op.op
+        src.children.discard(self)
+        del self  # TODO: why doesn't this delete remove it from the children
+        return src.permute_op(arg).reduce_op(rop, narg)
+
+      # move permutes before expands (always, this is safe)
+      if self.op.op == MovementOps.EXPAND:
+        self.op.src[0].children.discard(self)
+        return self.op.src[0].permute_op(arg).expand_op(tuple([self.op.arg[a] for a in arg]))
+
+      # move permutes before reshapes if we can
+      if PUSH_PERMUTES and self.op.op == MovementOps.RESHAPE and self.op.src[0].__class__ == LazyBuffer:
+        if shape_idx_groups := get_contraction(self.op.src[0].shape, self.shape):
+          self.op.src[0].children.discard(self)   # this changes nothing?
+          return self.op.src[0].permute_op(tuple(flatten(shape_idx_groups[i] for i in arg))) \
+            .reshape_op(ShapeTracker(self.st).permute(arg).shape)
+      if SHUFFLE_MOVEMENT_OPS and self.optype == BinaryOps and not self.realized and len(self.children) == 0:
+        return self.op.replace_with_movement_ops([(MovementOps.PERMUTE, arg)])
+    ret = create_lazybuffer(self.device, ShapeTracker(self.st).permute(arg), MovementOps, LazyOp(MovementOps.PERMUTE, (self,), arg, (self,)), self.dtype)
+    return self.prune_movement_nops(ret)
+  
+  def shrink_op(self:LazyBuffer, arg:Tuple[int, ...]) -> LazyBuffer:
+    if all([b - a == s for s, (a, b) in zip(self.shape, arg)]): return self
+    if not self.realized and self.op.op == MovementOps.SHRINK:
+      self.op.src[0].children.discard(self)
+      return self.op.src[0].shrink_op(tuple([(b1+b2, b1+e2) for (b1,e1),(b2,e2) in zip(self.op.arg, arg)]))
+    if SHUFFLE_MOVEMENT_OPS and self.optype == BinaryOps and not self.realized and len(self.children) == 0:
```

Comment:
Hmm, a lot of copy pasted code in these. I'm never willing to trade off maintainability/readability for a couple percent of speed.

Reply:
hmm here I can extract the `SHUFFLE_MOVEMENT_OPS` optimization into a helper (similar to what i did with `REMOVE_MOVEMENT_NOPS` and the `prune_movement_nops` helper)

Reply:
Yea, overall this change is +144 lines. I'm okay with lines, but not if it's copy/paste harder to read.

What would make this code the most readable? I'm okay with losing a few percent of perf.